org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGenerator.java

```diff
@@ -107,23 +138,26 @@ public Collection<MavenProjectInfo> createArchetypeProjects(IPath location, IArc
     userProperties.put("outputDirectory", basedir.getAbsolutePath());
     String projectFolder = location.append(artifactId).toFile().getAbsolutePath();
     File emptyPom = getEmptyPom(basedir);
+
+    CompletableFuture<?> mavenRun = null;
     try {
       String goals = "-U " + ArchetypePlugin.ARCHETYPE_PREFIX + ":generate";
       if(emptyPom != null) {
         goals += " -f " + emptyPom.getAbsolutePath();
       }
-      CompletableFuture<?> maven = mavenLauncher.runMaven(basedir, goals, userProperties, interactive);
+      mavenRun = mavenLauncher.runMaven(basedir, goals, userProperties, interactive);
       subMonitor.worked(1);
       Display current = Display.getCurrent();
-      while(!maven.isDone()) {
+      while(!mavenRun.isDone()) {
         if(current != null) {
           while(!current.isDisposed() && current.readAndDispatch()) {
             //loop to process events
```

Comment:
Yes, it will spend quite a bit in the UI thread pumping if the user is moving the mouse around hovering over things etc. I felt it safer, for an edge case of sorts, to keep the cancellation check outside.
IPM implementations SHOULD be smart about being asked to check cancellation very frequently. 
However I cannot 100% ascertain what implementation backs all uses of this code, and feel at least checking at anytime the event thread has an empty queue (alt-tab away to another program, or pause your mouse movement) , will catch the real bug fix: "something is wrong, I need control of eclipse back again". As such, I wouldn't do it.